File: ObjcExample/TabViewController.m
//
//  TabViewController.m
//  ObjcExample
//
//  Created by Timur Ganiev on 08.12.2020.
//  Copyright © 2020 Timur Ganiev. All rights reserved.
//

#import "TabViewController.h"

#import <TBTabBarController/TBTabBarControllerFramework.h>

#import "SettingsViewController.h"

@implementation TabViewController

#pragma mark - Public

#pragma mark Lifecycle

- (instancetype)init {
    
    if (@available(iOS 13.0, *)) {
        self = [super initWithStyle:UITableViewStyleInsetGrouped];
    } else {
        self = [super initWithStyle:UITableViewStyleGrouped];
    }
    
    return self;
}

#pragma mark Overrides

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    self.navigationController.navigationBar.prefersLargeTitles = true;
    self.navigationItem.searchController = [[UISearchController alloc] initWithSearchResultsController:nil];
}

- (void)viewDidAppear:(BOOL)animated {
    
    [super viewDidAppear:animated];
    
    self.navigationController.tb_tabBarItem.showsNotificationIndicator = false;
}

#pragma mark UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return 50;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"cell"];
    }
    
    NSString *title = [NSString stringWithFormat:@"Cell at index: %ld", indexPath.row];
    NSString *subtitle = @"Tap to show settings";
    
    if (@available(iOS 14.0, *)) {
        UIListContentConfiguration *config = [UIListContentConfiguration subtitleCellConfiguration];
        config.text = title;
        config.secondaryText = subtitle;
        cell.contentConfiguration = config;
    } else {
        cell.textLabel.text = title;
        cell.detailTextLabel.text = subtitle;
    }
    
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    
    return cell;
}

#pragma mark UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [tableView deselectRowAtIndexPath:indexPath animated:true];
    
    SettingsViewController *settingsViewController = [[SettingsViewController alloc] init];
    settingsViewController.tb_hidesTabBarWhenPushed = true;
    
    [self.navigationController pushViewController:settingsViewController animated:true];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    return 50.0;
}

@end

File: SwiftExample/TabViewController.swift
//
//  TabViewController.swift
//  SwiftExample
//
//  Created by Timur Ganiev on 09.12.2020.
//  Copyright © 2020-2023 Timur Ganiev. All rights reserved.
//

import UIKit
import TBTabBarControllerFramework

final class TabViewController: UITableViewController {

    // MARK: Lifecycle

    init() {
        if #available(iOS 13.0, *) {
            super.init(style: .insetGrouped)
        } else {
            super.init(style: .grouped)
        }
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { nil }

    // MARK: Overrides

    override func viewDidLoad() {
        super.viewDidLoad()
        setup()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        navigationController?.tb_tabBarItem.showsNotificationIndicator = false;
    }

    // MARK: UITableViewDataSource

    override func tableView(
        _ tableView: UITableView,
        numberOfRowsInSection section: Int
    ) -> Int {
        
        return 50
    }

    override func tableView(
        _ tableView: UITableView,
        cellForRowAt indexPath: IndexPath
    ) -> UITableViewCell {

        let identifier = NSStringFromClass(UITableViewCell.self)
        let cell = tableView.dequeueReusableCell(withIdentifier: identifier, for: indexPath)
        
        let title = "Cell at index \(indexPath.row)"
        let subtitle = "Tap to show settings"
        
        if #available(iOS 14.0, *) {
            var config = UIListContentConfiguration.subtitleCell()
            config.text = title
            config.secondaryText = subtitle
            cell.contentConfiguration  = config
        } else {
            cell.textLabel?.text = title
            cell.detailTextLabel?.text = subtitle
        }
        
        cell.accessoryType = .disclosureIndicator
        
        return cell
    }

    // MARK: UITableViewDelegate

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        tableView.deselectRow(at: indexPath, animated: true)
        
        let settingsViewController = SettingsViewController()
        settingsViewController.tb_hidesTabBarWhenPushed = true
        
        navigationController?.pushViewController(settingsViewController, animated: true)
    }

    override func tableView(
        _ tableView: UITableView,
        heightForRowAt indexPath: IndexPath
    ) -> CGFloat {
        
        return 50.0
    }
}

// MARK: Private Methods

private extension TabViewController {

    func setup() {

        navigationController?.navigationBar.prefersLargeTitles = true
        navigationItem.searchController = UISearchController(searchResultsController: nil)

        tableView.register(
            UITableViewCell.self,
            forCellReuseIdentifier: NSStringFromClass(UITableViewCell.self)
        )
    }
}
